File: 1813-maximum-erasure-value/1813-maximum-erasure-value.cpp
class Solution {
public:
    int maximumUniqueSubarray(vector<int>& nums) {
        int n = nums.size();
        unordered_map<int, int> mp;

        int i = 0;
        int j = 0;
        int ans = INT_MIN;

        int sum = 0;
        while(j < n) {
            mp[nums[j]]++;
            while(mp[nums[j]] > 1) {
                mp[nums[i]]--;
                sum -= nums[i];
                i++;
            }
            sum += nums[j];
            ans = max(sum, ans);
            j++;
        }

        return ans;
    }
};

File: 1813-maximum-erasure-value/1813-maximum-erasure-value.java
import java.util.HashMap;

class Solution {
    public int maximumUniqueSubarray(int[] nums) {
        int n = nums.length;
        HashMap<Integer, Integer> map = new HashMap<>();

        int i = 0, j = 0;
        int sum = 0;
        int ans = Integer.MIN_VALUE;

        while (j < n) {
            map.put(nums[j], map.getOrDefault(nums[j], 0) + 1);
            
            while (map.get(nums[j]) > 1) {
                map.put(nums[i], map.get(nums[i]) - 1);
                sum -= nums[i];
                i++;
            }

            sum += nums[j];
            ans = Math.max(ans, sum);
            j++;
        }

        return ans;
    }
}
